### PR TITLE
Add new IndexLogEntryTags to cache InMemoryFileIndex

### LINKED ISSUE #317: Cache InMemoryFileIndex for Index
**Feature requested**
image
```
val location =
          new InMemoryFileIndex(spark, index.content.files, Map(), None)
```
Listing files sometimes can take longer(3~4s).
There's `FileStatusCache` in Spark InMemoryFileIndex.scala, but we don't utilize it for Index. (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala)

We could use `FileStatusCache` of Spark if possible or create a custom cache for this.

**Acceptance criteria** 
- [x] listing index files performance should be improved when the cache is used.
- [x] InMemoryFileIndex in Hybrid Scan should also be handled properly

**Success criteria**

**Additional context**

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: #317 
 - **Parent Issue**: n/a
 - **Dependencies**: n/a
Fixes #317 

### What changes were proposed in this pull request?

This PR introduces 3 new IndexLogEntryTags for caching InMemoryFileIndex.
Currently, InMemoryFileIndex is created for every query execution. However, it incurs a new spark job & sometimes takes longer in cluster mode. To avoid the overhead, we could cache InMemoryFileIndex object for each index entry.

```scala
  // INMEMORYFILEINDEX_INDEX_ONLY stores InMemoryFileIndex for index only scan.
  val INMEMORYFILEINDEX_INDEX_ONLY: IndexLogEntryTag[InMemoryFileIndex] =
    IndexLogEntryTag[InMemoryFileIndex]("inMemoryFileIndexIndexOnly")

  // INMEMORYFILEINDEX_HYBRID_SCAN stores InMemoryFileIndex including index data files and also
  // appended files for Hybrid Scan.
  val INMEMORYFILEINDEX_HYBRID_SCAN: IndexLogEntryTag[InMemoryFileIndex] =
    IndexLogEntryTag[InMemoryFileIndex]("inMemoryFileIndexHybridScan")

  // INMEMORYFILEINDEX_HYBRID_SCAN_APPENDED stores InMemoryFileIndex including only appended files
  // for Hybrid Scan.
  val INMEMORYFILEINDEX_HYBRID_SCAN_APPENDED: IndexLogEntryTag[InMemoryFileIndex] =
    IndexLogEntryTag[InMemoryFileIndex]("inMemoryFileIndexHybridScanAppended")
```

#### Test result with cache

Test scripts:
```
hs.refreshIndex(indexName) // use refreshIndex to clear the tags.
val linetable = spark.read.parquet(tableName)
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count)
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count) // second query with cache

hs.refreshIndex(indexName)
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count)
hs.refreshIndex(indexName) // clear cache
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count) // second query without cache
```

Result:
```
linetable: org.apache.spark.sql.DataFrame = [l_orderkey: bigint, l_partkey: bigint ... 14 more fields]
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 2212
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 868
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 2235
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 2121
```

Spark UI:
image

### Does this PR introduce _any_ user-facing change?

Yes, if the cached InMemoryFileIndex object is used, we could avoid unnecessary listing files jobs for every query execution.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by imback82** [comment]
> Currently, InMemoryFileIndex is created for every query execution. However, it incurs a new spark job & sometimes takes longer in cluster mode.

Can you attach the screenshots of jobs from Spark UI if available (before and after)? It will be very useful to check the improvements.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 1]
```diff
@@ -450,14 +462,16 @@ object RuleUtils {
           }
           .getOrElse(Map())
 
-        val newLocation = new InMemoryFileIndex(spark, filesAppended, options, None)
+        val newLocation = index.getTagValueOrUpdate(originalPlan,
+          IndexLogEntryTags.INMEMORYFILEINDEX_HYBRID_SCAN_APPENDED,
+          new InMemoryFileIndex(spark, filesAppended, options, None))
```
[the remark this answers is not in the document]
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:467]: yea because hybrid scan tags are tagged with the plan.

[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:467]: Cool. Can you add a test? It will be the opposite of what you already added; check the cached file index is not used if plan changes.

**Comment by sezruby** [comment]
Test scripts:
```
hs.refreshIndex(indexName) // use refreshIndex to clear the tags.
val linetable = spark.read.parquet(tableName)
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count)
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count) // second query with cache

hs.refreshIndex(indexName)
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count)
hs.refreshIndex(indexName) // clear cache
val filter = linetable.filter(linetable("l_partkey") isin (1234,12341234, 123456)).select("l_suppkey","l_quantity","l_shipdate","l_extendedprice","l_discount","l_orderkey")
measure(filter.count) // second query without cache
```

Result:
```
linetable: org.apache.spark.sql.DataFrame = [l_orderkey: bigint, l_partkey: bigint ... 14 more fields]
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 2212
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 868
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 2235
filter: org.apache.spark.sql.DataFrame = [l_suppkey: bigint, l_quantity: double ... 4 more fields]
duration: 2121
```

Spark UI:
image

**Comment by imback82** [comment]
Great! Can you update the description with this info (you can just copy/paste)? Thanks!

**Comment by imback82** [comment]
@sezruby Can you fix the conflicts please?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 2]
```diff
@@ -382,7 +384,16 @@ object RuleUtils {
             baseRelation.schema.contains(s) || (filesDeleted.nonEmpty && s.name.equals(
               IndexConstants.DATA_FILE_NAME_ID))))
 
-        val newLocation = new InMemoryFileIndex(spark, filesToRead, Map(), None)
+        val newLocation = if (filesToRead.length == index.content.files.size) {
+          index.withCachedTag(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY) {
+            new InMemoryFileIndex(spark, index.content.files, Map(), None)
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:389]: Can `index.content.files` be replaced with `filesToRead`? In that case, we can just change tag based on the condition (easier to follow).

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 3]
```diff
@@ -474,14 +485,18 @@ object RuleUtils {
           }
           .getOrElse(Map())
 
-        val newLocation = new InMemoryFileIndex(spark, filesAppended, options, None)
+        val newLocation = index.withCachedTag(
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:488]: Seems much better with `withCachedTag` 👍🏼

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala [hunk 4]
```diff
@@ -670,13 +670,245 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
...
+      import spark.implicits._
+      SampleData.testData
+        .toDF("c1", "c2", "c3", "c4", "c5")
+        .limit(10)
+        .write
+        .parquet(testPath)
+      val df = spark.read.load(testPath)
+      hyperspace.createIndex(df, indexConfig)
+      withIndex(indexConfig.indexName) {
+
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala:697]:
```suggestion
```

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala [hunk 5]
```diff
@@ -670,13 +670,245 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
...
+        .json(testPath)
+      val df = spark.read.json(testPath)
+      hyperspace.createIndex(df, indexConfig)
+      withIndex(indexConfig.indexName) {
+        spark.enableHyperspace()
+        val df = spark.read.json(testPath)
+        def query(df: DataFrame): DataFrame = {
+          df.filter("c3 == 'facebook'").select("c3", "c4")
+        }
+        def getQueryPlanKey(df: DataFrame): LogicalPlan = {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala:831]: What do you mean "Key" here?

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala [hunk 6]
```diff
@@ -670,13 +670,245 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
...
+              .nonEmpty)
+          val location2 = getFsLocation(query(df2).queryExecution.optimizedPlan)
+          assert(!location2.eq(location1))
+
+          query(df2).collect
+          val location3 = getFsLocation(query(df2).queryExecution.optimizedPlan)
+          assert(location3.eq(location2))
+          assert(
+            location3.eq(indexEntry
+              .getTagValue(getQueryPlanKey(df2), IndexLogEntryTags.INMEMORYFILEINDEX_HYBRID_SCAN)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala:791]: nit: since we are using `IndexLogEntryTags` a lot, you can just `import com.microsoft.hyperspace.index.IndexLogEntryTags._` so that you don't have to specify IndexLogEntryTags

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala [hunk 7]
```diff
@@ -670,13 +670,245 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
...
+        assert(indexEntry.getTagValue(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY).nonEmpty)
+
+        SampleData.testData
+          .toDF("c1", "c2", "c3", "c4", "c5")
+          .limit(3)
+          .write
+          .mode("append")
+          .parquet(testPath)
+
+        val df2 = spark.read.parquet(testPath)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala:776]: Can you add short comments describing each test scenario? A bit hard to follow...

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala [hunk 8]
```diff
@@ -670,13 +670,245 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
...
+          spark.disableHyperspace()
+          val p = query(df).queryExecution.optimizedPlan
+          spark.enableHyperspace()
+          p
+        }
+        def getFsLocation(plan: LogicalPlan): FileIndex = {
+          plan.collect {
+            case LogicalRelation(HadoopFsRelation(loc, _, _, _, _, _), _, _, _) =>
+              loc
+          }.head
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala:756]: Any chance of reusing these across tests?

**Review by imback82 — COMMENTED** [review]
LGTM (few minor comments), thanks @sezruby!

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala (hunk 5 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala]: Renamed getOriginalQueryPlan; used "Key" because the plan before transformation is one of hash keys for IndexLogEntryTags

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 9]
```diff
@@ -382,7 +384,14 @@ object RuleUtils {
...
               IndexConstants.DATA_FILE_NAME_ID))))
 
-        val newLocation = new InMemoryFileIndex(spark, filesToRead, Map(), None)
+        def fileIndex: InMemoryFileIndex =
+          new InMemoryFileIndex(spark, filesToRead, Map(), None)
+        val newLocation = if (filesToRead.length == index.content.files.size) {
+          index.withCachedTag(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY)(fileIndex)
+        } else {
+          index.withCachedTag(plan, IndexLogEntryTags.INMEMORYFILEINDEX_HYBRID_SCAN)(fileIndex)
+        }
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:393]: I was thinking the following to make the intention clear, but the current approach is also fine:
```scala
        val tagForFileIndex = if (filesToRead.length == index.content.files.size) {
          IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY
        } else {
          IndexLogEntryTags.INMEMORYFILEINDEX_HYBRID_SCAN
        }
        val newLocation = index.withCachedTag(plan, tagForFileIndex) {
          new InMemoryFileIndex(spark, filesToRead, Map(), None)
        }
```

File: src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala [hunk 10]
```diff
@@ -808,4 +1017,22 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
...
+
+  private def getOriginalQueryPlan(query: DataFrame => DataFrame, df: DataFrame): LogicalPlan = {
+    spark.disableHyperspace()
+    val p = query(df).queryExecution.optimizedPlan
+    spark.enableHyperspace()
+    p
+  }
+
+  private def equalsRef(a: Set[FileIndex], b: Set[FileIndex]): Boolean = {
+    a.zip(b).forall(f => f._1 eq f._2)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala:1029]: Don't you need `a.size == b.size &&`?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 9 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:393]: it's because INDEX_ONLY doesn't require plan

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:393]: missed that. thanks

**Review by imback82 — APPROVED** [review]
